R/course.R

```diff
@@ -0,0 +1,324 @@
+    ## TO DO: it pains me that can_overwrite() always strips to basename
+    stop("Aborting download", call. = FALSE)
+  }
+
+  done(
+    "Downloading ZIP file to ",
+    if (is.null(destdir)) value(base_name) else value(full_path)
+  )
+  writeBin(dl$content, full_path)
+  invisible(full_path)
+}
+
+#' Unpack a ZIP file
+#'
+#' Special-purpose function to unpack a ZIP file and (attempt to) create the
+#' directory structure most people actually want. The main reason to finesse the
+#' directory structure: we want the same local result when unzipping the same
+#' content from either GitHub or DropBox ZIP files, which pack things
+#' differently. Here is the intent:
+#' * If the ZIP archive `foo.zip` does not contain a single top-level directory,
+#' i.e. it is packed as "loose parts", unzip into a directory named `foo`.
+#' Typical of DropBox ZIP files.
+#' * If the ZIP archive `foo.zip` has a single top-level directory (which, by
+#' the way, is not necessarily called "foo"), accept default unzip behavior.
+#' Typical of GitHub ZIP files.
+#'
+#' @section DropBox:
+#' The ZIP files produced by DropBox are special. The file list tends to contain
+#' a spurious directory `"/"`, which we ignore during unzip. Also, if the
+#' directory is a Git repo and/or RStudio Project, we unzip-ignore various
+#' hidden files, such as those below `.git/` and `.Rproj.user`.
+#'
+#' @param zipfile Path to local ZIP file.
+#'
+#' @return Path to the directory holding the unpacked files.
+#' @keywords internal
+#' @family download functions
+#' @export
+#' @examples
+#' \dontrun{
+#' download_zip("https://github.com/r-lib/rematch2/archive/master.zip")
+#' tidy_unzip("rematch2-master.zip")
+#' }
+tidy_unzip <- function(zipfile) {
+  filenames <- utils::unzip(zipfile, list = TRUE)[["Name"]]
+
+  ## deal with DropBox's peculiar habit of including "/" as a file --> drop it
+  filenames <- filenames[filenames != "/"]
+
+  ## DropBox ZIP files often include lots of hidden R, RStudio, and Git files
+  unzip_keep <- vapply(filenames, keep, logical(1), USE.NAMES = FALSE)
+  filenames <- filenames[unzip_keep]
+
+  td <- top_directory(filenames)
+  loose_parts <- is.na(td)
+
+  if (loose_parts) {
+    target <- tools::file_path_sans_ext(zipfile)
+    utils::unzip(zipfile, files = filenames, exdir = target)
+  } else {
+    target <- file.path(dirname(zipfile), td)
+    utils::unzip(zipfile, files = filenames, exdir = dirname(zipfile))
+  }
+  done(
+    "Unpacking ZIP file into ", value(target),
+    " (", length(filenames), " files extracted)"
+  )
+
+  if (interactive() &&
+      yep("Shall we delete the ZIP file ", value(zipfile), "?")) {
+    done("Deleting ", value(zipfile))
+    unlink(zipfile)
+  }
+
+  if (interactive()) {
+    done("Opening ", value(target), " in the file manager")
+    utils::browseURL(target)
+  }
+  invisible(target)
+}
+
+keep <- function(file,
+                 ignores = c(".Rproj.user", ".rproj.user", ".Rhistory", ".RData", ".git")) {
```

Comment:
Also, is my unzip-ignoring of `.RData` going to be an ugly surprise for anyone? It's the only here that I wondered about.

Reply:
At least, I've mentioned in the docs now.